cmake/Subdirectories.cmake

```diff
@@ -153,24 +160,108 @@ endfunction()
 #! A fatal error is logged indicating that is not gem could not be found in the list of external subdirectories
 function(query_gem_paths_from_external_subdirs output_gem_dirs gem_names registered_external_subdirs)
     if (gem_names)
-        foreach(gem_name IN LISTS gem_names)
+        foreach(gem_name_with_version_specifier IN LISTS gem_names)
             unset(gem_path)
+
+            # Remove the version specifier from the gem name before fetching properties
+            o3de_get_name_and_version_specifier(${gem_name_with_version_specifier} gem_name spec_op spec_version)
+
             get_property(gem_optional GLOBAL PROPERTY ${gem_name}_OPTIONAL)
-            o3de_find_gem_with_registered_external_subdirs(${gem_name} gem_path "${registered_external_subdirs}")
+            get_property(gem_path GLOBAL PROPERTY "@GEMROOT:${gem_name}@")
+            if(NOT gem_path)
+                if(NOT O3DE_DISABLE_GEM_DEPENDENCY_RESOLUTION)
+                    message(WARNING "Failed to find resolved gem path for gem '${gem_name_with_version_specifier}'")
+                    # Make a best effort attempt to find the most compatible gem
+                    o3de_find_most_compatible_gem_with_registered_external_subdirs(${gem_name_with_version_specifier} gem_path "${registered_external_subdirs}")
+                else()
+                    o3de_find_gem_with_registered_external_subdirs(${gem_name_with_version_specifier} gem_path "${registered_external_subdirs}")
+                endif()
+
+                if(gem_path)
+                    set_property(GLOBAL PROPERTY "@GEMROOT:${gem_name}@" "${gem_path}")
```

Comment:
Yes, before these changes, `add_o3de_object_json_external_subdirectories` would re-open all the `gem.json` files to determine the gem paths, so I left the logic to resolve those gem paths in that location, but if there is a better place to resolve these dependencies and set those gem paths, please let me know.  It's very likely I do not fully understand the gem dependency cmake logic.

Reply:
We can set each gems `@GEMROOT:<gem-name>@` property in `get_external_subdirectories_in_use` after all the external_subdirectories in use has been gathered.

If we have a mapping of "candidate gem_path to gem_name" that we populate when we `gem.json` file for the first time, then that would make sure we capture gems of the same name but different versions.
Then at the bottom of `get_external_subdirectories_in_use` e can loop over the "external subdirectories that will be used by the project" and use them as a key into the "candiadate_gem_path to gem_name" map to create mapping of `@GEMROOT@:<gem-name>` -> `<gem-path>`

Reply:
I spent some time trying this change and discovered that `query_gem_paths_from_external_subdirs` gets called for gems in the `external_subdirectories` in `engine.json` that are not used by the project or engine.  If we don't store a gem_name -> gem_path mapping like we are with `@GEMROOT:,gem-name>@` then we run logic that searches every external_subdirectory looking in all the gem.json files for a gem with that name.  None of those gem_name -> gem_path mappings will exist if we rely on the resolved gem_name -> gem_path mappings because the logic is looking for gems that the project doesn't use.  I think the reasoning here is that, even if your project doesn't use all the gems in the engine, it should still show them in the Visual Studio solution?

For example: The SaveData and GameStateSamples gems appear in Visual Studio even though they are not enabled in AutomatedTesting.